# X3FL dynamic relationships

## The relationship

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

In [2]:
all_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates", "Terran", "Yaki"]
# https://imgur.com/nqI0nbO, https://imgur.com/7BhbUMa, by blazenclaw, forum.egosoft.com
try:
    relationship_df = pd.read_csv("relationship.csv", index_col=0)
except:
    relationship_df = pd.read_csv("https://raw.githubusercontent.com/mkmark/X3FL-dynamic-relationships/main/relationship.csv", index_col=0)
relationship_df = pd.DataFrame(relationship_df)
cmap = sns.diverging_palette(5, 250, as_cmap=True)

def color_map(val):
    if val > 0.3:
        val = 0.3
    if val< -0.3:
        val = -0.3
    val = val/0.45+0.5
    cmap_val = cmap(val)
    colorvalue = (int(255*cmap_val[0]),int(255*cmap_val[1]),int(255*cmap_val[2]),cmap_val[3])
    color = 'rgba%s' % str(colorvalue)
    return 'background-color: %s' % color

relationship_df.style.applymap(color_map, subset=pd.IndexSlice[all_races, all_races])\
    .set_properties(**{'width': '40px', 'height': '40px'})\
    .set_caption("Relationship")\
    .set_precision(2)

The problem is simple: given the relationship $\mathbf{R}$, target notority $\mathbf{N}$, find tactics $\mathbf{X}$, which is a vector that represents how much notority point one should get from a race.

$$
\mathbf{R}\mathbf{X}=\mathbf{N}
$$

$$
\mathbf{X}=\mathbf{R}^{-1}\mathbf{N}
$$

Let's first define $\mathbf{N}$ as vectors of 1, since we don't want to lose effort by hitting the notoriety ceiling.

In [3]:
def get_X(races):
    R = np.array(relationship_df.loc[races, races])
    invR = np.linalg.inv(R) 
    N = np.ones(len(races))
    X = invR@N
    return X

def get_X_df(races):
    X = get_X(races)
    return pd.DataFrame(list(X)+[sum(X)]+[len(races)/sum(X)], index=races+['sum', 'efficiency']).transpose()

These races selected are the control group of races that are friends, but not necessarily to be the only friends as an notoriety overflow is allowed, which will make $N_i > 1$

To calculate the possible solutions, one has to recuse through all subsets and check if some get possible solutions, see below

## Example solutions

Let's see if there is a solution for all the races

The number for each race represents how much effort you will have to put with that race to get 1 notoriety point for each race assuming you're to be friend with all these selected race.

The sum is the total workload of current tactic.

The efficiency is the sum of actual gained notoriety points (equal the number of races selected) divided by total workload.

In [4]:
get_X_df(all_races)

Teladi      NMMC     Goner  TerraCorp  Strong Arms     Argon     Boron  \
0  0.413665  0.834924  0.400697  -0.602183    -0.088362 -0.559179 -0.534538   

     Split    Atreus      OTAS    Duke's   Paranid   Pirates    Terran  \
0 -0.60492 -0.931595 -1.156118 -0.343422 -1.250485 -1.373975 -1.513874   

       Yaki       sum  efficiency  
0 -1.536987 -8.846352   -1.695614

Unfortunately there is not a solution as one would have to get negative points from races. The problem is, when actively losing points of a race, the notoriety points of enemy races does not increase accordingly. However if that is true, we can see a path that you do some missions for some races but keep robbing other races and one day you will find yourself master of diplomacy with all races - too good to be true for those races you robbed.

Let's get results of some interesting races combo.

In [5]:
commonwealth_races = ["Teladi", "Argon", "Boron", "Split", "Paranid"]
major_races = ["Teladi", "Argon", "Boron", "Split", "Paranid", "Terran"]
all_but_yaki_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates", "Terran"]
all_but_yaki_terran_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates"]
all_but_yaki_terran_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid"]
all_but_yaki_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Terran"]
all_but_terran_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Yaki"]
all_but_teladi_yaki_terran_pirates_races = ["NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid"]

All commonwealth races, no problem

In [6]:
get_X_df(commonwealth_races)

Teladi     Argon     Boron     Split   Paranid       sum  efficiency
0     1.0  1.428571  1.428571  1.428571  1.428571  6.714286    0.744681

All major races, no problem

In [7]:
get_X_df(major_races)

Teladi     Argon     Boron     Split   Paranid    Terran        sum  \
0     1.0  3.823529  3.823529  3.823529  3.823529  5.588235  21.882353   

   efficiency  
0    0.274194

All but Yaki, not an option

In [8]:
get_X_df(all_but_yaki_races)

Teladi      NMMC    Goner  TerraCorp  Strong Arms     Argon     Boron  \
0  0.904032  0.639788  0.11671  -0.844907    -0.910454 -0.926158 -1.022721   

      Split   Atreus      OTAS    Duke's   Paranid  Pirates    Terran  \
0 -0.939952 -1.63911 -1.890735 -0.748691 -1.995335 -2.18678 -2.677981   

         sum  efficiency  
0 -14.122294    -0.99134

According to calculation just 3 enemies seems not sufficient

In [9]:
get_X_df(all_but_yaki_terran_pirates_races)

Teladi      NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0 -0.172497  7.816643  7.790471   13.40212    12.541359  10.368289  14.311163   

       Split     Atreus       OTAS     Duke's    Paranid         sum  \
0  17.984757  15.620415  18.506517  22.635896  21.638802  162.443937   

   efficiency  
0    0.073872

which is actually not the case since a subset is sufficient, just teledi notoriety points got overflowed doesn't render it illegal

In [10]:
get_X_df(all_but_teladi_yaki_terran_pirates_races)

NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0  7.780909  7.781132  13.390187    12.529926  10.355119  14.294295   

       Split     Atreus       OTAS    Duke's    Paranid         sum  \
0  17.964491  15.606104  18.489653  22.60303  21.615287  162.410133   

   efficiency  
0     0.06773

In [11]:
R = np.array(relationship_df.loc[all_races, all_but_teladi_yaki_terran_pirates_races])
R@get_X(all_but_teladi_yaki_terran_pirates_races)

array([  1.16713635,   1.        ,   1.        ,   1.        ,
         1.        ,   1.        ,   1.        ,   1.        ,
         1.        ,   1.        ,   1.        ,   1.        ,
       -24.73473903, -40.03737145, -40.29544981])

thus our complete solution

In [12]:
# https://stackoverflow.com/questions/26332412/python-recursive-function-to-display-all-subsets-of-given-set
def subs(l):
    if l == []:
        return [[]]

    x = subs(l[1:])

    return x + [[l[0]] + y for y in x]

def get_X_optimal_sub(races):
    possible_solution_racess = []
    possible_solution_Xs = []
    possible_solution_workloads = []
    for sub_races in subs(races):
        X = get_X(sub_races)
        R_all = np.array(relationship_df.loc[races, sub_races])
        N_all = R_all@X
        min_N_all = min(N_all)
        if min_N_all>0:
            if min_N_all<1:
                X = [x*1/min_N_all for x in X]
            possible_solution_racess += [sub_races]
            possible_solution_Xs += [X]
            possible_solution_workloads += [sum(X)]
    if len(possible_solution_workloads)>0:
        min_index = possible_solution_workloads.index(min(possible_solution_workloads))
        optiaml_solution_races = possible_solution_racess[min_index]
        optiaml_solution_X = possible_solution_Xs[min_index]
        if min(optiaml_solution_X)>0:
            return pd.DataFrame(list(optiaml_solution_X)+[sum(optiaml_solution_X)]+[len(races)/sum(optiaml_solution_X)], index=optiaml_solution_races+['sum', 'efficiency']).transpose()

In [13]:
get_X_optimal_sub(all_but_yaki_terran_pirates_races)

NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0  7.780909  7.781132  13.390187    12.529926  10.355119  14.294295   

       Split     Atreus       OTAS    Duke's    Paranid         sum  \
0  17.964491  15.606104  18.489653  22.60303  21.615287  162.410133   

   efficiency  
0    0.073887

This is actually a very interesting discovery that shows the best strategy to achieve the current relationship, is not just to improving relationship with every one but the enemies, i.e. Yaki, Terran, and Pirates. It also suggests you that you **should not improve relationship with Teladi**, which is quite contrary to intuition.

The logic behind this strategy is that by improving relationship with Teladi it leads to loss of relationship in general. By improving relationship with Teladi's ally, it would already make notoriety points of more than enough to achieve good relationship with Teladi, as verified above.

## Let's explore all possibilities

Turned out there are the following number of possible combos.

In [14]:
least_enemy_set_list = []
for sub_races in subs(all_races):
    if sub_races != []:
        X = get_X(sub_races)
        if min(X)>0:
            R_all = np.array(relationship_df.loc[all_races, sub_races])
            N_all = R_all@X
            enemy_set = set([race for i, race in enumerate(all_races) if N_all[i]<=0])
            if enemy_set == set():
                continue
            is_duplicate = False
            for i in reversed(range(len(least_enemy_set_list))):
                least_enemy_set = least_enemy_set_list[i]
                if least_enemy_set.issubset(enemy_set):
                    is_duplicate = True
                    break
                if enemy_set.issubset(least_enemy_set):
                    least_enemy_set_list.pop(i)
            if not is_duplicate:
                least_enemy_set_list = least_enemy_set_list + [enemy_set]
                
len(least_enemy_set_list)

283

which is the following

In [15]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    workload = sum(get_X(list(set(all_races) - least_enemy_set)))
    efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
    item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
    print(item)
    parsed_least_enemy_set_list += [item]

Boron - TerraCorp - Strong Arms - Split - Paranid - Pirates - NMMC - Argon - Teladi, 1758.3749999999743, 0.0034122414160802374
Atreus - TerraCorp - Split - Terran - Pirates - OTAS - NMMC - Strong Arms - Teladi, 544.480076202567, 0.011019686967880483
Atreus - TerraCorp - Split - Paranid - Pirates - OTAS - NMMC - Argon - Teladi, 552.8193767538032, 0.010853454586256453
Atreus - TerraCorp - Split - Terran - OTAS - NMMC - Argon - Teladi, 506.48866444611303, 0.013820644945045464
Teladi - TerraCorp - Terran - Pirates - OTAS - NMMC - Argon - Atreus, 15209.636542237855, 0.00046023453489902146
Boron - Teladi - TerraCorp - Pirates - OTAS - NMMC - Terran - Atreus, 618.5248369292437, 0.011317249659290183
Teladi - TerraCorp - Paranid - Pirates - OTAS - NMMC - Terran - Atreus, 11356.756358877454, 0.0006163731772345522
Boron - Atreus - Split - Pirates - OTAS - Argon - NMMC - Strong Arms - Teladi, 479.08273381294765, 0.01252393287532385
Terran - Paranid - Pirates - OTAS - Argon - NMMC - Strong Arms - T

Boron - TerraCorp - Duke's - Paranid - OTAS - Terran, 120.68555809056849, 0.07457396015226568
TerraCorp - Split - Duke's - Paranid - Pirates - Atreus, 497.05351949403536, 0.018106702089467852
TerraCorp - Split - Paranid - OTAS - NMMC - Terran, 2505.2585002528967, 0.003592443653655494
TerraCorp - Split - Duke's - Pirates - Terran, 633.29026127524, 0.015790547575867127
TerraCorp - Split - Paranid - Pirates - NMMC - Yaki, 1932.8745328669559, 0.004656277397711199
TerraCorp - Split - NMMC - Terran - Yaki, 5771.820407590587, 0.0017325556399587357
TerraCorp - Duke's - Paranid - OTAS - Terran - Atreus, 73.66057299503643, 0.12218205254263308
TerraCorp - Duke's - Pirates - Terran - Atreus, 93.52379886462637, 0.10692465577103832
TerraCorp - Duke's - Pirates - OTAS - Terran, 114.82559841822973, 0.0870885946840613
TerraCorp - Duke's - Paranid - Pirates - Terran, 304.5529752978415, 0.032835010034692226
Boron - Split - Duke's - Pirates - OTAS - Argon - Strong Arms, 108.11800528520361, 0.0739932260024

3 or less enemies option are listed below

In [16]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) <= 3:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

Terran - Yaki - Paranid, 195.74987462921928, 0.061302721254508426
Pirates - Terran - Yaki, 162.44393680858647, 0.07387163987622407


4 enemies

In [17]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 4:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

TerraCorp - Yaki - Argon - Terran, 6380.36023270709, 0.0017240405868639906
Boron - TerraCorp - Yaki - Terran, 1167.9560442782533, 0.009418162655939272
Yaki - TerraCorp - Atreus - Terran, 174.70372019814613, 0.06296374220036058
OTAS - TerraCorp - Yaki - Terran, 178.81625696398214, 0.06151565963163887
Duke's - TerraCorp - Yaki - Terran, 1169.8407341546097, 0.009402989380387063
OTAS - Strong Arms - Yaki - Terran, 281.97836679426337, 0.039010084798546986
OTAS - Argon - Yaki - Terran, 83.24190571176665, 0.13214498041513598
Boron - Terran - Atreus - Yaki, 362.03269977700705, 0.030383995718550886
Boron - OTAS - Terran - Yaki, 78.34237671512548, 0.14040932201992082
OTAS - Terran - Yaki - Split, 121.09694482534431, 0.09083631313626515
Yaki - OTAS - Terran - Atreus, 151.9081199563972, 0.07241219233808814
Yaki - Duke's - Terran - Atreus, 134248.85468439595, 8.19373843140768e-05
OTAS - Duke's - Terran - Yaki, 68.01331847910076, 0.16173302885346047
OTAS - NMMC - Terran - Yaki, 824.8106996204912, 0.

4 enemies but without terran

In [18]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 4 and "Terran" not in least_enemy_set:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

5 enemies but without any major races

In [19]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 5 and set(major_races).intersection(least_enemy_set) == set():
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

Duke's - Pirates - Yaki - Strong Arms - Atreus, 115.80946764695189, 0.08634872608589528
Duke's - Pirates - OTAS - Strong Arms - Yaki, 55.18972548571196, 0.1811931462240901
TerraCorp - Duke's - Pirates - Yaki - Atreus, 280.24981341472966, 0.035682450161711364
Duke's - Pirates - OTAS - Yaki - Atreus, 59.131288178803295, 0.16911520631449198


enemies including Yaki has the following number of possible combos

In [20]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if "Yaki" in least_enemy_set:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        # print(item)
        parsed_least_enemy_set_list += [item]
        
len(parsed_least_enemy_set_list)

116

## Conclusion

- A simple tool is developed
- All 283 possible combos are given with total workload and efficiency
- The only two options to maintain at most 3 enemies (races, workload, efficiency)
    - Terran - Yaki - Paranid, 195.74987462921916, 0.06130272125450846
    - Pirates - Terran - Yaki, 162.4439368085863, 0.07387163987622415
- 4 options to keep good relationship with all 6 major races with least enemies (5) are listed as below:
    - Atreus - Pirates - Duke's - Strong Arms - Yaki, 115.80946764695183, 0.08634872608589532
    - OTAS - Pirates - Duke's - Strong Arms - Yaki, 55.18972548571193, 0.18119314622409022
    - Atreus - TerraCorp - Pirates - Duke's - Yaki, 280.2498134147285, 0.03568245016171151
    - OTAS - Atreus - Pirates - Duke's - Yaki, 59.131288178803246, 0.16911520631449212